File: 0501-0600/0566.reshape-the-matrix.cpp
class Solution {
public:
    vector<vector<int>> matrixReshape(vector<vector<int>>& nums, int r, int c) {
        const int R = nums.size();
        if (R == 0) return nums;
        const int C = nums[0].size();
        if (C == 0) return nums;
        if (R * C != r * c) return nums;
        vector<vector<int> > target(r, vector<int>(c));
        for (int i = 0; i < R * C; ++i) {
            target[i / c][i % c] = nums[i / C][i % C];
        }
        return target;
    }
};

File: 0501-0600/0566-reshape-the-matrix.rs
struct Solution {}

impl Solution {
    pub fn matrix_reshape(nums: Vec<Vec<i32>>, r: i32, c: i32) -> Vec<Vec<i32>> {
        let old_r = nums.len();
        if old_r == 0 {
            return nums;
        }
        let old_c = nums[0].len();
        let r = r as usize;
        let c = c as usize;
        if old_c == 0 || r * c != old_r * old_c {
            return nums;
        }
        nums.into_iter().flatten().collect::<Vec<i32>>().
            chunks(c).map(|x| x.to_vec()).collect()
    }
}

fn main() {
    let nums = vec![vec![1, 2, 3, 4], vec![5, 6, 7, 8], vec![9, 10, 11, 12]];
    println!("{:?}", Solution::matrix_reshape(nums, 2, 6));
}
